exercises/practice/collatz-conjecture/.approaches/match-recursion/content.md

```diff
@@ -0,0 +1,109 @@
+object CollatzConjecture {
+
+  private object Status extends Enumeration {
+    type Status = Value
+    val ILLEGAL, EVEN, ODD, ONE = Value
+  }
+  import Status._
+
+  private def getStatus(num: Integer): Status = {
+    if (num <= 0)
+      ILLEGAL
+    else if (num == 1)
+      ONE
+    else if (num % 2 == 0)
+      EVEN
+    else
+      ODD
+  }
+
+  @tailrec
+  private def collatzMeBaby(steps: Int, num: Int): Option[Int] = {
+    getStatus(num) match {
+      case ILLEGAL => None
+      case ONE     => Some(steps)
+      case EVEN    => collatzMeBaby(steps + 1, num / 2)
+      case ODD     => collatzMeBaby(steps + 1, (num * 3) + 1)
+    }
+  }
+
+  def steps(start: Int) = {
+    collatzMeBaby(0, start)
+  }
+}
+```
+
+This approach starts by importing from packages for what will be needed.
+
+The `CollatzConjecture` class starts by defining an [Enumeration][enum-v2] in the version 2 way.
+With version 3, an [`enum`](https://docs.scala-lang.org/scala3/reference/enums/enums.html) can be created like so
+
+```scala
+enum Status:
+  case ILLEGAL, EVEN, ODD, ONE
+```
+
+The `getStatus()` method returns an `enum` value depending on the value of the passed-in `Int`.
+
+The `collatzMeBaby()` method is annotated with the [`@tailrec`][tailrec-annotation] annotation to verify that the method can be compiled
+with [tail call optimization][tail-opt].
+
+A tail call is a particular form of [recursion][recursion] where the last call in the method is a call to the same method _and nothing else_.
+
+In other words, if the last call in `recurMe()` is `recurMe(arg1, arg2) + 1`, the `+ 1` makes the recursion non-tail recursive.
+
+If the last call in `recurMe()` is `recurMe(arg1, arg2, acc + 1)`, then the recursion is a tail call, because only the method is being called
+with no other operation being peformed on it.
+
+A [`match`][match] expression is used to perform [pattern matching][pattern-matching] on the result of passing in the number to
+the `getStatus()` method.
+
+If the `Status` of the number is `ILLEGAL`, then `None` is returned.
+If the `Status` of the number is `ONE`, then the number of steps wrapped in `Some` is returned from the function.
+
+If the `Status` of the number is `EVEN` or `ODD`, then the function calls itself by adding `1` to `steps` and calculating
+the new value for the number according to whether it is even or odd.
+
+The `steps()` method calls `collatzMeBaby()`, passing in `0` as the starting value for `steps` and the number passed into the `steps()`
+method as the number to be recalculated until it reaches `1`.
+
+## Shortening
+
+All of the ceremony involving the enumeration could be gotten rid of by using [pattern guards][pattern-guard], and the code shortened to
+
+```scala
+import scala.annotation.tailrec
+
+object CollatzConjecture {
+
+  @tailrec
+  private def collatzMeBaby(steps: Int, num: Int): Option[Int] = {
+    num match {
+      case nbr0 if nbr0 <= 0     => None
+      case num1 if num1 == 1     => Some(steps)
```

Comment:
Could this be simplified to:

```suggestion
      case 1     => Some(steps)
```

Reply:
Yes it can! But the formatting of the edit is off a bit, so I'll edit it directly. Thanks!